skxray/core/speckle.py

```diff
@@ -0,0 +1,288 @@
+
+    start_time = time.time()  # used to log the computation time (optionally)
+
+    for i, images in enumerate(image_sets):
+        # Ring buffer, a buffer with periodic boundary conditions.
+        # Images must be keep for up to maximum delay in buf.
+        buf = np.zeros([num_times, timebin_num],
+                       dtype=np.object)  # matrix of buffers
+
+        # to track processing each time level
+        track_level = np.zeros(num_times)
+
+        # to increment buffer
+        cur = np.ones(num_times)*timebin_num
+
+        # to track how many images processed in each level
+        img_per_level = np.zeros(num_times, dtype=np.int64)
+
+        prob_k = np.zeros_like(prob_k_all)
+        prob_k_pow = np.zeros_like(prob_k_all)
+
+        for n, img in enumerate(images):
+            cur[0] = (1 + cur[0]) % timebin_num
+            # read each frame
+            # Put the image into the ring buffer.
+            buf[0, cur[0] - 1] = (np.ravel(img))[indices]
+
+            _process(num_roi, 0, cur[0] - 1, buf, img_per_level, labels,
+                     max_cts, bin_edges[0, 0], prob_k, prob_k_pow)
+
+            # check whether the number of levels is one, otherwise
+            # continue processing the next level
+            processing = num_times > 1
+            level = 1
+
+            while processing:
+                if not track_level[level]:
+                    track_level[level] = 1
+                    processing = 0
+                else:
+                    prev = 1 + (cur[level - 1] - 2) % timebin_num
+                    cur[level] = 1 + cur[level] % timebin_num
+
+                    buf[level, cur[level]-1] = (buf[level-1,
+                                                    prev-1] +
+                                                buf[level-1,
+                                                    cur[level - 1] - 1])
+                    track_level[level] = 0
+
+                    _process(num_roi, level, cur[level]-1, buf, img_per_level,
+                             labels, max_cts, bin_edges[level, 0], prob_k,
+                             prob_k_pow)
+                    level += 1
+                    # Checking whether there is next level for processing
+                    processing = level < num_times
+
+            prob_k_all += (prob_k - prob_k_all)/(i + 1)
+            prob_k_pow_all += (prob_k_pow - prob_k_pow_all)/(i + 1)
+
+    prob_k_std_dev = np.power((prob_k_pow_all -
+                               np.power(prob_k_all, 2)), .5)
+
+    logger.info("Processing time for XSVS took %s seconds."
+                "", (time.time() - start_time))
+    return prob_k_all, prob_k_std_dev
+
+
+def _process(num_roi, level, buf_no, buf, img_per_level, labels,
+             max_cts, bin_edges, prob_k, prob_k_pow):
+    """
+    Internal helper function. This modifies inputs in place.
+
+    This helper function calculate probability of detecting speckles for
+    each integration time.
```

Comment:
Can you copy [this pattern] (https://github.com/scikit-xray/scikit-xray/blob/05201515ce7746619420f078b3114bba7f9428ab/skxray/core/fitting/xrf_model.py#L164 ) and add a `..warning : This function mutates the input values`?